cajita/src/Cajita_DynamicPartitioner.hpp

```diff
@@ -903,6 +762,225 @@ class SparseDimPartitioner : public BlockPartitioner<NumSpaceDim>
+        MPI_Comm comm, float max_workload_coeff, int workload_num,
+        int num_step_rebalance,
+        const std::array<int, num_space_dim>& ranks_per_dim,
+        const std::array<int, num_space_dim>& global_cells_per_dim,
+        int max_optimize_iteration = 10 )
+        : base( comm, max_workload_coeff, workload_num, num_step_rebalance,
+                ranks_per_dim, global_cells_per_dim, max_optimize_iteration )
+    {
+    }
+
+    /*!
+      \brief compute the workload in the current MPI rank from particle
+      positions (each particle count for 1 workload value). This function must
+      be called before running optimizePartition() \param view particle
+      positions view \param particle_num total particle number \param
+      global_lower_corner the coordinate of the domain global lower corner
+      \param dx cell dx size
+      \param comm MPI communicator used for workload reduction
+    */
+    template <class ParticlePosViewType, typename ArrayType, typename CellUnit>
+    void setLocalWorkloadByParticles( const ParticlePosViewType& view,
+                                      int particle_num,
+                                      const ArrayType& global_lower_corner,
+                                      const CellUnit dx, MPI_Comm comm )
+    {
+        base::resetWorkload();
+        // make a local copy
+        auto workload = _workload_per_tile;
+        Kokkos::Array<CellUnit, num_space_dim> lower_corner;
+        for ( std::size_t d = 0; d < num_space_dim; ++d )
+        {
+            lower_corner[d] = global_lower_corner[d];
+        }
+
+        Kokkos::parallel_for(
+            "compute_local_workload_parpos",
+            Kokkos::RangePolicy<execution_space>( 0, particle_num ),
+            KOKKOS_LAMBDA( const int i ) {
+                int ti = static_cast<int>(
+                             ( view( i, 0 ) - lower_corner[0] ) / dx - 0.5 ) >>
+                         cell_bits_per_tile_dim;
+                int tj = static_cast<int>(
+                             ( view( i, 1 ) - lower_corner[1] ) / dx - 0.5 ) >>
+                         cell_bits_per_tile_dim;
+                int tz = static_cast<int>(
+                             ( view( i, 2 ) - lower_corner[2] ) / dx - 0.5 ) >>
+                         cell_bits_per_tile_dim;
+                Kokkos::atomic_increment( &workload( ti + 1, tj + 1, tz + 1 ) );
+            } );
+        Kokkos::fence();
+        // Wait for other ranks' workload to be ready
+        MPI_Barrier( comm );
+    }
+};
+
+/*!
+  Dynamic mesh block partitioner. (Current Version: Support 3D only) Workload
+  are computed from sparse map occupancy.
+
+  \tparam Device Kokkos device type.
+  \tparam CellPerTileDim Cells per tile per dimension.
+  \tparam NumSpaceDim Dimemsion (The current version support 3D only)
+*/
+template <typename Device, unsigned long long CellPerTileDim = 4,
+          std::size_t NumSpaceDim = 3>
+class SparseMapDynamicPartitioner
```

Comment:
Let's move the sparse version into a separate `Cajita_SparseDynamicPartitioner.hpp`